src/net_processing.cpp

```diff
@@ -281,6 +303,180 @@ struct Peer {
 
         /** Minimum fee rate with which to filter transaction announcements to this node. See BIP133. */
         std::atomic<CAmount> m_fee_filter_received{0};
+
+        void AddKnownTx(const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(!m_tx_inventory_mutex)
+        {
+            AssertLockNotHeld(m_tx_inventory_mutex);
+
+            LOCK(m_tx_inventory_mutex);
+            m_tx_inventory_known_filter.insert(hash);
+        }
+
+        void RelayTransaction(const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(!m_tx_inventory_mutex)
+        {
+            AssertLockNotHeld(m_tx_inventory_mutex);
+            LOCK(m_tx_inventory_mutex);
+            if (!m_tx_inventory_known_filter.contains(hash)) {
+                m_tx_inventory_to_send.insert(hash);
+            }
+        }
+
+        bool VerifyUnconfirmedParent(const uint256& parent_txid) EXCLUSIVE_LOCKS_REQUIRED(!m_tx_inventory_mutex)
+        {
+            AssertLockNotHeld(m_tx_inventory_mutex);
+            return WITH_LOCK(m_tx_inventory_mutex, return !m_tx_inventory_known_filter.contains(parent_txid));
+        }
+
+        void SetSendMempool(bool value) EXCLUSIVE_LOCKS_REQUIRED(!m_tx_inventory_mutex)
+        {
+            AssertLockNotHeld(m_tx_inventory_mutex);
+            LOCK(m_tx_inventory_mutex);
+            m_send_mempool = value;
+        }
+
+        void Relay(
```

Comment:
I haven't reviewed this pull request, but this would be the first time that a net processing logic is put in the peers struct, instead of the net processing implementation.

Reply:
The idea is to prevent external synchronizing (acquiring `m_tx_inventory_mutex` from outside of the `CNode::TxRelay` instance).
Perhaps this requires a more complex refactoring.

Reply:
@dergoegge you might have thoughts here?